File: app/src/main/java/org/stepic/droid/storage/dao/ProgressDaoImpl.java
package org.stepic.droid.storage.dao;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import org.stepic.droid.model.Progress;
import org.stepic.droid.storage.structure.DbStructureProgress;

import javax.inject.Inject;

public class ProgressDaoImpl extends DaoBase<Progress> {


    @Inject
    public ProgressDaoImpl(SQLiteDatabase openHelper) {
        super(openHelper);
    }

    @Override
    public Progress parsePersistentObject(Cursor cursor) {
        Progress progress = new Progress();
        int indexId = cursor.getColumnIndex(DbStructureProgress.Column.ID);
        int indexCost = cursor.getColumnIndex(DbStructureProgress.Column.COST);
        int indexScore = cursor.getColumnIndex(DbStructureProgress.Column.SCORE);
        int indexIs_Passed = cursor.getColumnIndex(DbStructureProgress.Column.IS_PASSED);
        int indexLastViewed = cursor.getColumnIndex(DbStructureProgress.Column.LAST_VIEWED);
        int indexSteps = cursor.getColumnIndex(DbStructureProgress.Column.N_STEPS);
        int indexN_steps_passed = cursor.getColumnIndex(DbStructureProgress.Column.N_STEPS_PASSED);

        progress.setId(cursor.getString(indexId));
        progress.setCost(cursor.getInt(indexCost));
        progress.setScore(cursor.getString(indexScore));
        progress.setPassed(cursor.getInt(indexIs_Passed) > 0);
        progress.setLastViewed(cursor.getString(indexLastViewed));
        progress.setNSteps(cursor.getInt(indexSteps));
        progress.setNStepsPassed(cursor.getInt(indexN_steps_passed));

        return progress;
    }

    @Override
    public String getDbName() {
        return DbStructureProgress.PROGRESS;
    }

    @Override
    public ContentValues getContentValues(Progress progress) {

        ContentValues values = new ContentValues();
        values.put(DbStructureProgress.Column.ID, progress.getId());
        values.put(DbStructureProgress.Column.COST, progress.getCost());
        values.put(DbStructureProgress.Column.SCORE, progress.getScore());
        values.put(DbStructureProgress.Column.IS_PASSED, progress.isPassed());
        values.put(DbStructureProgress.Column.LAST_VIEWED, progress.getLastViewed());
        values.put(DbStructureProgress.Column.N_STEPS, progress.getNSteps());
        values.put(DbStructureProgress.Column.N_STEPS_PASSED, progress.getNStepsPassed());
        return values;
    }

    @Override
    public String getDefaultPrimaryColumn() {
        return DbStructureProgress.Column.ID;
    }

    @Override
    public String getDefaultPrimaryValue(Progress persistentObject) {
        return persistentObject.getId();
    }

}

File: app/src/main/java/org/stepic/droid/storage/dao/ProgressDaoImpl.kt
package org.stepic.droid.storage.dao

import android.content.ContentValues
import android.database.Cursor
import org.stepic.droid.storage.operations.DatabaseOperations
import org.stepic.droid.storage.structure.DbStructureProgress
import org.stepic.droid.util.getBoolean
import org.stepic.droid.util.getLong
import org.stepic.droid.util.getString
import org.stepik.android.model.Progress
import javax.inject.Inject

class ProgressDaoImpl
@Inject
constructor(
    databaseOperations: DatabaseOperations
) : DaoBase<Progress>(databaseOperations) {
    public override fun getDbName(): String =
        DbStructureProgress.TABLE_NAME

    public override fun getDefaultPrimaryColumn(): String =
        DbStructureProgress.Columns.ID

    public override fun getDefaultPrimaryValue(persistentObject: Progress): String? =
        persistentObject.id

    public override fun parsePersistentObject(cursor: Cursor): Progress =
        Progress(
            id      = cursor.getString(DbStructureProgress.Columns.ID).orEmpty(),
            cost    = cursor.getLong(DbStructureProgress.Columns.COST),
            score   = cursor.getString(DbStructureProgress.Columns.SCORE),
            isPassed     = cursor.getBoolean(DbStructureProgress.Columns.IS_PASSED),
            lastViewed   = cursor.getString(DbStructureProgress.Columns.LAST_VIEWED),
            nStepsPassed = cursor.getLong(DbStructureProgress.Columns.N_STEPS_PASSED),
            nSteps       = cursor.getLong(DbStructureProgress.Columns.N_STEPS)
        )

    public override fun getContentValues(progress: Progress): ContentValues {
        val values = ContentValues()
        values.put(DbStructureProgress.Columns.ID, progress.id)
        values.put(DbStructureProgress.Columns.COST, progress.cost)
        values.put(DbStructureProgress.Columns.SCORE, progress.score)
        values.put(DbStructureProgress.Columns.IS_PASSED, progress.isPassed)
        values.put(DbStructureProgress.Columns.LAST_VIEWED, progress.lastViewed)
        values.put(DbStructureProgress.Columns.N_STEPS, progress.nSteps)
        values.put(DbStructureProgress.Columns.N_STEPS_PASSED, progress.nStepsPassed)
        return values
    }
}
